File: src/main/java/com/oycm/algorithm/da/advance/Solution_8.scala
package com.oycm.algorithm.da.advance

object Solution_8 {

  /**
   * 2070. 每一个查询的最大美丽值 1724
   * https://leetcode.cn/problems/most-beautiful-item-for-each-query/description/
   *
   * 求出价格小于等于 queries(i) items 中的最大美丽值
   *
   * @param items   items(i) = [price, beauty] 二维数组
   * @param queries 查询数组
   * @return
   */
  def maximumBeauty(items: Array[Array[Int]], queries: Array[Int]): Array[Int] = {
    /*
    先对 items 按价格升序排序，再从左到右 遍历出每个 items(i)(1) 的最大值（每个价格能获得的最大值）
    找 小于等于 queries(i) 价格最大的美丽值，就是找到 大于等于 queries(i) + 1 的 index - 1
     */
    val n = items.length
    // 排序
    val sorted = items.sortBy(row => row(0))
    var max = 0
    for (elem <- sorted) {
      max = Math.max(elem(1), max);
      elem(1) = max
    }

    def lowerBound(nums: Array[Array[Int]], x: Int): Int = {
      var l = -1
      var r = nums.length
      while (l + 1 < r) {
        val mid = l + (r - l) / 2
        if (nums(mid)(0) >= x) {
          r = mid
        } else {
          l = mid
        }
      }

      r
    }

    for (i <- queries.indices) {
      val x = queries(i)
      var beauty = 0
      if (x < sorted(0)(0)) {
        beauty = 0
      } else if (x >= sorted(n - 1)(0)) {
        beauty = sorted(n - 1)(1)
      } else {
        beauty = sorted(lowerBound(sorted, x + 1) - 1)(1)
      }
      queries(i) = beauty
    }

    queries;
  }

  def main(args: Array[String]): Unit = {
    println(maximumBeauty(Array(Array(1, 2), Array(3, 2), Array(2, 4), Array(5, 6), Array(3, 5)), Array(1, 2, 3, 4, 5, 6)).toSeq)
  }

}

File: src/main/java/com/oycm/algorithm/g/diff_one_basic/Solution_8.java
package com.oycm.algorithm.g.diff_one_basic;

public class Solution_8 {

    /**
     * 3355. <a href="https://leetcode.cn/problems/zero-array-transformation-i/description/">零数组变换 I</a> 1591
     *
     * @param nums    长度为 n 的整数数组, [1, 1e5]
     * @param queries queries[i] = [l, r]
     * @return
     */
    public boolean isZeroArray(int[] nums, int[][] queries) {
        /*
        每个查询 queries[i]
            在 nums 的下标范围 [l, r] 内选择一个下标 子集(区间中任意挑选一些元素操作)
            将选中的每个下标对应的元素值减 1
        零数组：指所有元素都等于 0 的数组
        如果在按顺序处理所有查询后，可以将 nums 转换为 零数组 ，则返回 true，否则返回 false
         */
        /*
        将 nums 解析成一个差分数组，对差分数组进行查询操作，对差分数组进行还原，
            如果所以元素都小于等于 0 ,则返回 true；
            否则，返回 false
         */
        int n = nums.length;
        int[] diff = new int[n];
        diff[0] = nums[0];
        for (int i = 1; i < n; i++) {
            diff[i] = nums[i] - nums[i - 1];
        }
        for (int[] query : queries) {
            diff[query[0]] -= 1;
            if (query[1] < n - 1) {
                diff[query[1] + 1] += 1;
            }
        }
        if (diff[0] > 0) {
            return false;
        }
        for (int i = 1; i < n; i++) {
            diff[i] += diff[i - 1];
            if (diff[i] > 0) {
                return false;
            }
        }

        return true;
    }

}
